pkg/controller/csr/csr_controller.go

```diff
@@ -0,0 +1,344 @@
+func relevantCSR(csr *certificatesv1.CertificateSigningRequest) bool {
+	if csr.Spec.SignerName != certificatemanager.OperatorCSRSignerName {
+		return false
+	}
+	if _, found := csr.Labels[LabelName]; !found {
+		return false
+	}
+	for _, condition := range csr.Status.Conditions {
+		if (condition.Type == certificatesv1.CertificateDenied || condition.Type == certificatesv1.CertificateFailed) && condition.Status == corev1.ConditionTrue {
+			// These request statuses are deterministic/non-recoverable.
+			return false
+		}
+	}
+	// We are only interested in CSRs that have not been signed. This also excludes
+	// us from watching our own updates and reconciling for no reason.
+	return csr.Status.Certificate == nil
+}
+
+// Add creates a new CSR Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager, opts options.AddOptions) error {
+	if !opts.EnterpriseCRDExists {
+		// No need to start this controller (currently).
+		return nil
+	}
+
+	ctrl, err := controller.New(controllerName, mgr, controller.Options{Reconciler: newReconciler(mgr, opts)})
+	if err != nil {
+		return err
+	}
+
+	return utils.AddCSRWatchWithRelevancyFn(ctrl, relevantCSR)
+}
+
+type tlsAsset struct {
+	serviceaccountName      string
+	serviceaccountNamespace string
+	validDNSNames           []string
+}
+
+func newReconciler(mgr manager.Manager, opts options.AddOptions) reconcile.Reconciler {
+	r := &ReconcileCSR{
+		client:           mgr.GetClient(),
+		scheme:           mgr.GetScheme(),
+		provider:         opts.DetectedProvider,
+		clusterDomain:    opts.ClusterDomain,
+		allowedTLSAssets: allowedAssets(opts.ClusterDomain),
+	}
+	return r
+}
+
+// allowedAssets To prevent any abuse of this controller for obtaining a fraudulent certificate, this controller
+// will only approve a pre-defined list of assets, based on their 'requestor', dns names and namespaces.
+// Some of the information a CSR will contain:
+//   - Name: The name is based on the secret name + a pod suffix. We use the secret name as the key to index the map.
+//   - Requestor: this is the user identity tied to the request. This will be matched against the sa + namespace.
+//   - DNS names: these will be checked against pre-defined dns names for that specific secret name.
+//
+// The combination of this information (among other checks) will help us reject/approve requests.
+func allowedAssets(clusterDomain string) map[string]tlsAsset {
+	return map[string]tlsAsset{
+		rmonitor.PrometheusServerTLSSecretName: {
+			serviceaccountName:      rmonitor.PrometheusServiceAccountName,
+			serviceaccountNamespace: rmonitor.TigeraPrometheusObjectName,
+			validDNSNames:           monitor.PrometheusTLSServerDNSNames(clusterDomain),
+		},
+	}
+}
+
+// blank assignment to verify that ReconcileCompliance implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileCSR{}
```

Comment:
Here you're asserting that a pointer to this structure implements the interface, but the bound functions are all defined against the struct and not a pointer to the struct. I don't think it matters a ton, but we usually default to pointer receivers unless there's a specific reason not to so probably best to do that here as well.

`func (r ReconcileCSR)` -> `func (r *ReconcileCSR)`